packages/virtualized-lists/Lists/VirtualizedList.js

```diff
@@ -712,14 +747,43 @@ class VirtualizedList extends StateSafePureComponent<Props, State> {
       return prevState;
     }
 
+    let maintainVisibleContentPositionAdjustment =
+      prevState.maintainVisibleContentPositionAdjustment;
+    const prevFirstItemKey = prevState.firstItemKey;
+    const newFirstItemKey = VirtualizedList._getItemKey(newProps, 0);
+    if (
+      newProps.maintainVisibleContentPosition != null &&
+      maintainVisibleContentPositionAdjustment == null &&
+      prevFirstItemKey != null &&
+      newFirstItemKey != null
+    ) {
+      // Fast path if items were added at the start of the list.
+      const hint = itemCount - prevState.renderMask.numCells();
+      maintainVisibleContentPositionAdjustment =
+        newFirstItemKey !== prevFirstItemKey
+          ? VirtualizedList._findItemIndexWithKey(
+              newProps,
+              prevFirstItemKey,
+              hint > 0 ? hint : null,
+            )
+          : null;
+    }
+
     const constrainedCells = VirtualizedList._constrainToItemCount(
       prevState.cellsAroundViewport,
       newProps,
     );
 
+    if (maintainVisibleContentPositionAdjustment != null) {
+      constrainedCells.first += maintainVisibleContentPositionAdjustment;
+      constrainedCells.last += maintainVisibleContentPositionAdjustment;
+    }
+
     return {
       cellsAroundViewport: constrainedCells,
       renderMask: VirtualizedList._createRenderMask(newProps, constrainedCells),
+      firstItemKey: newFirstItemKey,
+      maintainVisibleContentPositionAdjustment,
```

Comment:
It seems like:
1. We never use the adjustment value outside of here, apart from as a check as to whether we have an updated scroll position yet
2. We do not use `firstItemKey` outside of this function.

So could we instead have simpler state like: `pendingScrollPositionUpdate: boolean`? Or `pendingScrollUpdateCount: number` (see below).

Reply:
We do need `firstItemKey` as state to check if the value changed.